pkg/controller/replica.go

```diff
@@ -86,74 +86,68 @@ func (r *ReconcileReplica) Reconcile(ctx context.Context, request reconcile.Requ
-							return reconcile.Result{Requeue: true}, err
-						}
-					}
-				}
-
-				if !isCleanupRequested(azVolumeAttachment) {
-					go r.handleReplicaDelete(context.Background(), azVolumeAttachment)
-				}
-			} else if azVolumeAttachment.Status.State == azdiskv1beta2.AttachmentFailed {
-				// if the failed attachment isn't retriable, delete the CRI so that replacing replica AzVolumeAttachment can be created
-				// once the attachment has ever been retriable, it should be retried until it succeeds or gets detached
-				if !azureutils.MapContains(azVolumeAttachment.Status.Annotations, consts.ReplicaVolumeAttachRetryAnnotation) && !azureutils.MapContains(azVolumeAttachment.Status.Annotations, consts.ReplicaVolumeAttachRetryCount) {
+			if azVolumeAttachment.Status.State == azdiskv1beta2.DetachmentFailed {
+				// if detachment failed and the driver is uninstalling, delete azvolumeattachment CRI to let uninstallation proceed
+				if r.driverLifecycle.IsDriverUninstall() {
+					r.logger.Info("Deleting AzVolumeAttachment in DetachmentFailed state without detaching disk", workflow.GetObjectDetails(azVolumeAttachment)...)
 					if err := r.cachedClient.Delete(ctx, azVolumeAttachment); err != nil {
 						return reconcile.Result{Requeue: true}, err
 					}
-					go r.handleReplicaDelete(context.Background(), azVolumeAttachment)
-					// else, reset the state to AttachmentPending to retry the attachment
 				} else {
-					retryCount := 0
-					if value, exists := azureutils.GetFromMap(azVolumeAttachment.Status.Annotations, consts.ReplicaVolumeAttachRetryCount); exists {
-						if n, err := strconv.Atoi(value); err == nil {
-							retryCount = n
-						} else {
-							r.logger.Error(err, "failed to convert retry count to int")
-						}
+					if _, err := azureutils.UpdateCRIWithRetry(ctx, nil, r.cachedClient, r.azClient, azVolumeAttachment, func(obj client.Object) error {
+						azVolumeAttachment := obj.(*azdiskv1beta2.AzVolumeAttachment)
+						_, err = updateState(azVolumeAttachment, azdiskv1beta2.ForceDetachPending, normalUpdate)
+						return err
+					}, consts.NormalUpdateMaxNetRetry, azureutils.UpdateCRIStatus); err != nil {
+						return reconcile.Result{Requeue: true}, err
 					}
-
-					var updateFunc azureutils.UpdateCRIFunc
-					if retryCount < r.config.ControllerConfig.ReplicaVolumeAttachRetryLimit {
-						updateFunc = func(obj client.Object) error {
-							azva := obj.(*azdiskv1beta2.AzVolumeAttachment)
-							azureutils.RemoveFromMap(azva.Status.Annotations, consts.ReplicaVolumeAttachRetryAnnotation)
-							azureutils.AddToMap(azva.Status.Annotations, consts.ReplicaVolumeAttachRetryCount, strconv.Itoa(retryCount+1))
-							_, derr := updateState(azva, azdiskv1beta2.AttachmentPending, forceUpdate)
-							azva.Status.Detail = nil
-							azva.Status.Error = nil
-							return derr
-						}
+				}
+			} else if !isCleanupRequested(azVolumeAttachment) {
```

Comment:
Now that there is no check for `objectDeletionRequested` at https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/0c0cbfc785a90526cc7beb51b04e7bb6e82d493a/pkg/controller/replica.go#LL89C55-L89C55, we should move this check to https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/0c0cbfc785a90526cc7beb51b04e7bb6e82d493a/pkg/controller/replica.go#LL111C4-L111C4